File: server/src/db/repositories/worktree.repository.ts
import Database from 'better-sqlite3'
import type { WorktreeRow, SortMode } from '@claude-manager/shared'
import { BaseRepository } from './base.repository.js'

export interface CreateWorktreeDto {
  workspaceId: string
  name: string
  branch: string
  path: string
  isMain?: boolean
}

export interface UpdateWorktreeDto {
  name?: string
  branch?: string
  sortMode?: SortMode
  displayOrder?: number
}

export class WorktreeRepository extends BaseRepository<WorktreeRow> {
  constructor(db: Database.Database) {
    super(db)
  }

  findById(id: string): WorktreeRow | null {
    const stmt = this.db.prepare('SELECT * FROM worktrees WHERE id = ?')
    return stmt.get(id) as WorktreeRow | null
  }

  findByPath(path: string): WorktreeRow | null {
    const stmt = this.db.prepare('SELECT * FROM worktrees WHERE path = ?')
    return stmt.get(path) as WorktreeRow | null
  }

  findByWorkspaceId(workspaceId: string): WorktreeRow[] {
    const stmt = this.db.prepare(`
      SELECT * FROM worktrees
      WHERE workspace_id = ?
      ORDER BY display_order ASC
    `)
    return stmt.all(workspaceId) as WorktreeRow[]
  }

  findMainWorktree(workspaceId: string): WorktreeRow | null {
    const stmt = this.db.prepare(`
      SELECT * FROM worktrees
      WHERE workspace_id = ? AND is_main = 1
    `)
    return stmt.get(workspaceId) as WorktreeRow | null
  }

  create(dto: CreateWorktreeDto): WorktreeRow {
    const id = this.generateId('wt')
    const now = this.now()

    // Get next display order
    const maxOrder = this.db
      .prepare(
        `
      SELECT MAX(display_order) as max FROM worktrees WHERE workspace_id = ?
    `
      )
      .get(dto.workspaceId) as { max: number | null }

    const displayOrder = (maxOrder.max ?? -1) + 1

    const stmt = this.db.prepare(`
      INSERT INTO worktrees (id, workspace_id, name, branch, path, display_order, is_main, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `)
    stmt.run(
      id,
      dto.workspaceId,
      dto.name,
      dto.branch,
      dto.path,
      displayOrder,
      dto.isMain ? 1 : 0,
      now,
      now
    )

    return this.findById(id)!
  }

  update(id: string, dto: UpdateWorktreeDto): WorktreeRow | null {
    const updates: string[] = []
    const params: unknown[] = []

    if (dto.name !== undefined) {
      updates.push('name = ?')
      params.push(dto.name)
    }
    if (dto.branch !== undefined) {
      updates.push('branch = ?')
      params.push(dto.branch)
    }
    if (dto.sortMode !== undefined) {
      updates.push('sort_mode = ?')
      params.push(dto.sortMode)
    }
    if (dto.displayOrder !== undefined) {
      updates.push('display_order = ?')
      params.push(dto.displayOrder)
    }

    if (updates.length === 0) {
      return this.findById(id)
    }

    updates.push('updated_at = ?')
    params.push(this.now())
    params.push(id)

    const stmt = this.db.prepare(`
      UPDATE worktrees
      SET ${updates.join(', ')}
      WHERE id = ?
    `)
    stmt.run(...params)

    return this.findById(id)
  }

  delete(id: string): boolean {
    const stmt = this.db.prepare('DELETE FROM worktrees WHERE id = ?')
    const result = stmt.run(id)
    return result.changes > 0
  }

  reorder(workspaceId: string, worktreeIds: string[]): void {
    const stmt = this.db.prepare(`
      UPDATE worktrees
      SET display_order = ?, updated_at = ?
      WHERE id = ? AND workspace_id = ?
    `)

    this.transaction(() => {
      const now = this.now()
      worktreeIds.forEach((id, index) => {
        stmt.run(index, now, id, workspaceId)
      })
    })
  }

  existsWithNameInWorkspace(workspaceId: string, name: string, excludeId?: string): boolean {
    let stmt
    if (excludeId) {
      stmt = this.db.prepare(`
        SELECT 1 FROM worktrees WHERE workspace_id = ? AND name = ? AND id != ? LIMIT 1
      `)
      return stmt.get(workspaceId, name, excludeId) !== undefined
    }
    stmt = this.db.prepare(`
      SELECT 1 FROM worktrees WHERE workspace_id = ? AND name = ? LIMIT 1
    `)
    return stmt.get(workspaceId, name) !== undefined
  }
}

File: src-tauri/src/db/repositories/worktree_repository.rs
//! Worktree repository for database operations

use rusqlite::params;

use crate::db::{DbPool, DbResult};
use crate::types::{Worktree, WorktreeRow};

pub struct WorktreeRepository {
    pool: DbPool,
}

impl WorktreeRepository {
    pub fn new(pool: DbPool) -> Self {
        Self { pool }
    }

    pub fn find_by_id(&self, id: &str) -> DbResult<Option<Worktree>> {
        let conn = self.pool.get()?;
        let mut stmt = conn.prepare(
            r#"
            SELECT id, workspace_id, name, branch, path, sort_mode, display_order, is_main, created_at, updated_at
            FROM worktrees WHERE id = ?
        "#,
        )?;

        let row = stmt
            .query_row([id], |row| {
                Ok(WorktreeRow {
                    id: row.get(0)?,
                    workspace_id: row.get(1)?,
                    name: row.get(2)?,
                    branch: row.get(3)?,
                    path: row.get(4)?,
                    sort_mode: row.get(5)?,
                    display_order: row.get(6)?,
                    is_main: row.get::<_, i32>(7)? != 0,
                    created_at: row.get(8)?,
                    updated_at: row.get(9)?,
                })
            })
            .optional()?;

        Ok(row.map(Worktree::from))
    }

    pub fn find_by_path(&self, path: &str) -> DbResult<Option<Worktree>> {
        let conn = self.pool.get()?;
        let mut stmt = conn.prepare(
            r#"
            SELECT id, workspace_id, name, branch, path, sort_mode, display_order, is_main, created_at, updated_at
            FROM worktrees WHERE path = ?
        "#,
        )?;

        let row = stmt
            .query_row([path], |row| {
                Ok(WorktreeRow {
                    id: row.get(0)?,
                    workspace_id: row.get(1)?,
                    name: row.get(2)?,
                    branch: row.get(3)?,
                    path: row.get(4)?,
                    sort_mode: row.get(5)?,
                    display_order: row.get(6)?,
                    is_main: row.get::<_, i32>(7)? != 0,
                    created_at: row.get(8)?,
                    updated_at: row.get(9)?,
                })
            })
            .optional()?;

        Ok(row.map(Worktree::from))
    }

    pub fn find_by_workspace_id(&self, workspace_id: &str) -> DbResult<Vec<Worktree>> {
        let conn = self.pool.get()?;
        let mut stmt = conn.prepare(
            r#"
            SELECT id, workspace_id, name, branch, path, sort_mode, display_order, is_main, created_at, updated_at
            FROM worktrees WHERE workspace_id = ? ORDER BY display_order, created_at
        "#,
        )?;

        let rows = stmt.query_map([workspace_id], |row| {
            Ok(WorktreeRow {
                id: row.get(0)?,
                workspace_id: row.get(1)?,
                name: row.get(2)?,
                branch: row.get(3)?,
                path: row.get(4)?,
                sort_mode: row.get(5)?,
                display_order: row.get(6)?,
                is_main: row.get::<_, i32>(7)? != 0,
                created_at: row.get(8)?,
                updated_at: row.get(9)?,
            })
        })?;

        let worktrees: Vec<Worktree> = rows.filter_map(|r| r.ok()).map(Worktree::from).collect();

        Ok(worktrees)
    }

    pub fn create(&self, worktree: &Worktree) -> DbResult<Worktree> {
        let conn = self.pool.get()?;

        conn.execute(
            r#"
            INSERT INTO worktrees (id, workspace_id, name, branch, path, sort_mode, display_order, is_main, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        "#,
            params![
                worktree.id,
                worktree.workspace_id,
                worktree.name,
                worktree.branch,
                worktree.path,
                worktree.sort_mode.as_str(),
                worktree.display_order,
                worktree.is_main as i32,
                worktree.created_at,
                worktree.updated_at,
            ],
        )?;

        self.find_by_id(&worktree.id)?
            .ok_or_else(|| rusqlite::Error::QueryReturnedNoRows.into())
    }

    pub fn update(&self, worktree: &Worktree) -> DbResult<Worktree> {
        let conn = self.pool.get()?;

        conn.execute(
            r#"
            UPDATE worktrees SET
                name = ?,
                branch = ?,
                sort_mode = ?,
                display_order = ?,
                updated_at = datetime('now')
            WHERE id = ?
        "#,
            params![
                worktree.name,
                worktree.branch,
                worktree.sort_mode.as_str(),
                worktree.display_order,
                worktree.id,
            ],
        )?;

        self.find_by_id(&worktree.id)?
            .ok_or_else(|| rusqlite::Error::QueryReturnedNoRows.into())
    }

    pub fn delete(&self, id: &str) -> DbResult<()> {
        let conn = self.pool.get()?;
        conn.execute("DELETE FROM worktrees WHERE id = ?", [id])?;
        Ok(())
    }

    pub fn reorder(&self, workspace_id: &str, worktree_ids: &[String]) -> DbResult<()> {
        let conn = self.pool.get()?;

        for (index, id) in worktree_ids.iter().enumerate() {
            conn.execute(
                r#"
                UPDATE worktrees SET display_order = ?, updated_at = datetime('now')
                WHERE id = ? AND workspace_id = ?
            "#,
                params![index as i32, id, workspace_id],
            )?;
        }

        Ok(())
    }
}

// Helper trait for optional query results
trait OptionalExt<T> {
    fn optional(self) -> Result<Option<T>, rusqlite::Error>;
}

impl<T> OptionalExt<T> for Result<T, rusqlite::Error> {
    fn optional(self) -> Result<Option<T>, rusqlite::Error> {
        match self {
            Ok(value) => Ok(Some(value)),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(e),
        }
    }
}
